File: web/lib/chunking.ts
// Dynamic-window Parent-Child chunking (port of chunking.py).
//
// Group Whisper segments into "parent" windows (break at a natural pause or
// after ~30s), then split each parent into "child" sentences. Only children are
// embedded, but every child carries its parent's wide time window so retrieval
// splices a complete thought rather than a fragment.

import { config } from "./config";
import type { Segment, Parent, ChildRecord } from "./types";

function splitSentences(text: string): string[] {
  const trimmed = text.trim();
  if (!trimmed) return [];
  // Split after sentence-ending punctuation followed by whitespace.
  return trimmed
    .split(/(?<=[.!?])\s+/)
    .map((s) => s.trim())
    .filter(Boolean);
}

export function buildParents(segments: Segment[]): Parent[] {
  const parents: Parent[] = [];
  let current: Parent | null = null;

  for (const seg of segments) {
    const text = (seg.text ?? "").trim();
    if (!text) continue;

    if (!current) {
      current = { start: seg.start, end: seg.end, text };
      continue;
    }

    const gap = seg.start - current.end;
    const duration = seg.end - current.start;
    if (gap > config.pauseThresholdSeconds || duration > config.parentMaxSeconds) {
      parents.push(current);
      current = { start: seg.start, end: seg.end, text };
    } else {
      current.end = seg.end;
      current.text = `${current.text} ${text}`.trim();
    }
  }

  if (current) parents.push(current);
  return parents;
}

export function buildChildRecords(
  parents: Parent[],
  filePath: string,
  fileId: string,
  audioType: string
): ChildRecord[] {
  const records: ChildRecord[] = [];
  parents.forEach((parent, pIdx) => {
    const startMs = Math.round(parent.start * 1000);
    const endMs = Math.round(parent.end * 1000);
    splitSentences(parent.text).forEach((sentence, cIdx) => {
      records.push({
        _id: `${fileId}-p${pIdx}-c${cIdx}`,
        file_path: filePath,
        file_id: fileId,
        child_text: sentence,
        parent_text: parent.text,
        start_time_ms: startMs,
        end_time_ms: endMs,
        audio_type: audioType,
      });
    });
  });
  return records;
}

File: chunking.py
"""Dynamic-window Parent-Child chunking.

Strategy (from the spec): preserve linear human speech and avoid micro-cuts.

- Group Whisper *segments* into "Parent" blocks. A parent breaks when there is
  a natural pause (gap > PAUSE_THRESHOLD) or it would exceed PARENT_MAX_SECONDS.
- Split each Parent into "Child" sentences. Only the child sentences are
  embedded, but every child carries its *parent's* wide time window so that
  retrieval splices a complete thought rather than a fragment.
"""

import re
from typing import List, Dict, Any

import config


def _split_sentences(text: str) -> List[str]:
    """Lightweight sentence splitter (no NLP model downloads)."""
    text = text.strip()
    if not text:
        return []
    parts = re.split(r"(?<=[.!?])\s+", text)
    return [p.strip() for p in parts if p.strip()]


def build_parents(segments: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """Group transcript segments into parent windows.

    Each segment is expected to have ``start`` and ``end`` (seconds) and ``text``.
    Returns parents with ``start``/``end`` (seconds) and concatenated ``text``.
    """
    parents: List[Dict[str, Any]] = []
    current: Dict[str, Any] = {}

    for seg in segments:
        start = float(seg["start"])
        end = float(seg["end"])
        text = (seg.get("text") or "").strip()
        if not text:
            continue

        if not current:
            current = {"start": start, "end": end, "text": text}
            continue

        gap = start - current["end"]
        duration = end - current["start"]
        if gap > config.PAUSE_THRESHOLD_SECONDS or duration > config.PARENT_MAX_SECONDS:
            parents.append(current)
            current = {"start": start, "end": end, "text": text}
        else:
            current["end"] = end
            current["text"] = f"{current['text']} {text}".strip()

    if current:
        parents.append(current)
    return parents


def build_child_records(
    parents: List[Dict[str, Any]],
    file_path: str,
    file_id: str,
    audio_type: str,
) -> List[Dict[str, Any]]:
    """Flatten parents into per-child Pinecone records following the spec schema."""
    records: List[Dict[str, Any]] = []
    for p_idx, parent in enumerate(parents):
        start_ms = int(round(parent["start"] * 1000))
        end_ms = int(round(parent["end"] * 1000))
        sentences = _split_sentences(parent["text"])
        for c_idx, sentence in enumerate(sentences):
            records.append(
                {
                    "_id": f"{file_id}-p{p_idx}-c{c_idx}",
                    "file_path": file_path,
                    "file_id": file_id,
                    "child_text": sentence,
                    "parent_text": parent["text"],
                    "start_time_ms": start_ms,
                    "end_time_ms": end_ms,
                    "audio_type": audio_type,
                }
            )
    return records
